src/include/migraphx/iterator.hpp

```diff
@@ -95,24 +97,28 @@ struct iterator_operators
     // Core operators
 
     // Increment
-    template <class U, MIGRAPHX_REQUIRES(std::is_same<U, T>{})>
-    friend constexpr auto operator++(U& x) -> decltype(U::increment(x), std::declval<U&>())
+    template <class U,
+              class Self = std::remove_reference_t<U>,
+              MIGRAPHX_REQUIRES(std::is_same<Self, T>{})>
+    friend constexpr auto operator++(U&& x) -> decltype(Self::increment(x), static_cast<U&&>(x))
     {
-        U::increment(x);
-        return x;
+        Self::increment(x);
```

Comment:
Self used so we can ensure reference isn't returned? Previous case ran into stale refs being used?

Reply:
The extra template is used to make sure it is SFINAE friendly.

Reply:
`Self` is added to remove the reference because I switched it from an lvalue reference to an universal reference(So `U` can now get deduced as a reference).  This is to allow the operator to work with temporaries like `++v.begin()` which there are tests for that.